rs-matter/src/data_model/system_model/acl.rs

```diff
@@ -15,226 +15,214 @@
-                    }
+        attr: &AttrDetails<'_>,
+        builder: ArrayAttributeRead<
+            AccessControlEntryStructArrayBuilder<P>,
+            AccessControlEntryStructBuilder<P>,
+        >,
+    ) -> Result<P, Error> {
+        let mut acls = fabric_mgr
+            .iter()
+            .filter(|fabric| !attr.fab_filter || fabric.fab_idx().get() == attr.fab_idx)
+            .flat_map(|fabric| fabric.acl_iter().map(|entry| (fabric.fab_idx(), entry)));
+
+        match builder {
+            ArrayAttributeRead::ReadAll(mut builder) => {
+                for (fab_idx, entry) in acls {
+                    builder = entry.read_into(fab_idx, builder.push()?)?;
                 }
+
+                builder.end()
+            }
+            ArrayAttributeRead::ReadOne(index, builder) => {
+                let Some((fab_idx, entry)) = acls.nth(index as usize) else {
+                    return Err(ErrorCode::InvalidAction.into()); // TODO
+                };
+
+                entry.read_into(fab_idx, builder)
             }
-        } else {
-            Ok(())
         }
     }
 
-    /// Write the ACL Attribute
-    ///
-    /// This takes care of 4 things, add item, edit item, delete item, delete list.
-    /// Care about fabric-scoped behaviour is taken
-    fn write_acl_attr(
+    /// For unit-testing
+    /// Set the ACL entries in the fabric manager
+    fn set_acl(
         &self,
         fabric_mgr: &mut FabricMgr,
-        op: &ListOperation,
-        data: &TLVElement,
         fab_idx: NonZeroU8,
+        value: ArrayAttributeWrite<
+            TLVArray<'_, AccessControlEntryStruct<'_>>,
+            AccessControlEntryStruct<'_>,
+        >,
+        _notify: &dyn ChangeNotify,
     ) -> Result<(), Error> {
-        debug!("Performing ACL operation {:?}", op);
-        match op {
-            ListOperation::AddItem | ListOperation::EditItem(_) => {
-                let acl_entry = AclEntry::from_tlv(data)?;
-                debug!("ACL  {:?}", acl_entry);
-
-                if let ListOperation::EditItem(index) = op {
-                    fabric_mgr.acl_update(fab_idx, *index as _, acl_entry)?;
-                } else {
-                    fabric_mgr.acl_add(fab_idx, acl_entry)?;
+        match value {
+            ArrayAttributeWrite::Replace(list) => {
+                // Check the well-formedness of the list first
+                for entry in &list {
+                    let entry = entry?;
+                    entry.check()?;
+                }
+                if list.iter().count() > acl::ENTRIES_PER_FABRIC {
+                    Err(ErrorCode::InvalidAction)?;
                 }
 
-                Ok(())
+                // Now add everything
+                fabric_mgr.acl_remove_all(fab_idx)?;
+                for entry in list {
+                    let entry = unwrap!(entry);
+                    unwrap!(fabric_mgr.acl_add_init(fab_idx, AclEntry::init_with(fab_idx, &entry)));
```

Comment:
Can/should we propagate the error instead of unwrapping here?

Reply:
I added a comment why the `unwrap!` calls should never fail because apparently it is not clear.

To some extent this code is a manual re-implementation of the generic functionality in draft PR #214 which I'm not pushing as of yet. If we decide in favor of #214, this code should get much simpler.